src/puppetlabs/jdbc_util/pool.clj

```diff
@@ -0,0 +1,95 @@
+(ns puppetlabs.jdbc-util.pool
+  (:require [clojure.set :as set]
+            [clojure.tools.logging :as log])
+  (:import [com.zaxxer.hikari HikariConfig HikariDataSource]
+           java.io.Closeable
+           java.sql.SQLTransientConnectionException
+           javax.sql.DataSource))
+
+(defn connection-pool
+  [^HikariConfig config]
+  (let [ds (HikariDataSource. config)]
+    {:datasource ds}))
+
+(defn- set-option
+  [^HikariConfig config option value]
+  (case option
+    :username (.setUsername config value)
+    :password (.setPassword config value)
+    :data-source-class-name (.setDataSourceClassName config value)
+    :jdbc-url (.setJdbcUrl config value)
+    :driver-class-name (.setDriverClassName config value)
+    :auto-commit (.setAutoCommit config value)
+    :connection-timeout (.setConnectionTimeout config value)
+    :idle-timeout (.setIdleTimeout config value)
+    :max-lifetime (.setMaxLifetime config value)
+    :connection-test-query (.setConnectionTestQuery config value)
+    :minimum-idle (.setMinimumIdle config value)
+    :maximum-pool-size (.setMaximumPoolSize config value)
+    :metric-registry (.setMetricRegistry config value)
+    :health-check-registry (.setHealthCheckRegistry config value)
+    :pool-name (.setPoolName config value)
+    (throw (IllegalArgumentException. (format "%s is not a recognized HikariCP option" (str option))))))
+
+(defn options->hikari-config
+  [options]
+  (let [config (HikariConfig.)]
+    (doseq [[option value] options]
+      (set-option config option value))
+    (.validate config)
+    config))
+
+(defn spec->hikari-options
+  [db-spec]
+  (-> db-spec
+      (set/rename-keys {:user :username
+                        :classname :driver-class-name})
+      (assoc :jdbc-url (str "jdbc:"
+                            (:subprotocol db-spec) ":"
+                            (:subname db-spec)))
+      (dissoc :subprotocol :subname)))
+
+(defn wrap-with-delayed-init
+  "Wraps a connection pool that loops trying to get a connection, and then runs
+  init-fn (with the connection as argument) before returning any connections to
+  the application. Accepts a timeout in ms that's used when deferencing the
+  future. The datasource should have initialization-fail-fast set before being
+  created or this is pointless."
+  [^DataSource datasource init-fn timeout]
+  (let [pool-future
+        (future
```

Comment:
it's unbounded, [created](https://github.com/clojure/clojure/blob/f5f4faf95051f794c9bfa0315e4457b600c84cef/src/jvm/clojure/lang/Agent.java#L53-L54) using [Executors.newCachedThreadPool](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool%28%29)